### DISCUSSION TITLE
Add configuration toggle for disabling in-memory cache for static pages

### DISCUSSION THREAD

**Discussion by pekkai-solteq** [vercel/next.js#26247] [Help] [answered]
### Summary

When deploying Next.js applications on top of Kubernetes, it quickly becomes apparent that the incremental static regeneration is not very easy to get working nicely. This is because the pods would individually write the page files inside of the container filesystem and the static pages go out of sync between the pods.

A way to reduce the problem is to have a shared build volume that is mounted inside of all of the pods, but even if this is done, the individual pods will still continue to use the in-memory cache for storing the pre-rendered json / html contents to some extent (see https://github.com/vercel/next.js/blob/canary/packages/next/next-server/server/incremental-cache.ts#L115-L122). In essence the filesystem doesn't really reflect the real html / json content that the client would receive.

### Proposed solution

Specifically for Kubernetes deployments, I propose that a new experimental `disableLruCache` toggle is added to the Next configuration. When enabled, this toggle should add the following behaviour:

* disable usage of LRUCache in `incremental-cache.ts` entirely
    * bypass reading of html / json data from the cache, and instead always read it from the filesystem
* calculate revalidation time based on 'last modified' date-time of a given pre-rendered file, if the file exists - something along the lines of:

```
  private calculateRevalidateTime(
    pathname: string,
    configuredRevalidate: number
  ): number {
    // if cache is disabled, let's calculate the validation time
    // based on when file was last modified
    if (!this.useLruCache) {
      try {
        const { mtimeMs } = statSync(this.getSeedPath(pathname, 'html'))
        return mtimeMs + configuredRevalidate * 1000
      } catch (e) {
        // file doesn't exist yet
      }
    }
    // fall back to default behaviour
    return configuredRevalidate * 1000 + new Date().getTime()
  }
```

This comes at a performance cost, but the reality is that the best practice would anyways be to run a Nginx (or some other) ingress service in front which would handle gzipping and caching of the static json / html content. 

### Implementation

In order to opt for this functionality, the configuration just needs to be changed to reflect this:

    module.exports = {
        experimental: {
            disableLruCache: true,
        },
    };

**Comment by sherifabdlnaby** [comment]
Ignoring issues like this one makes me question how Vercel is committed to making NextJS true OSS and not a "vercel-first" freemium product. (related: https://github.com/vercel/next.js/discussions/27687#discussioncomment-1121013) 

The lack of such a feature cripples NextJS in production once you need >1 server, and using Vercel is not always an option for some.

**Comment by angelo-hub** [comment]
I want to and likely will open a PR for this, less ignorable. Without this feature it severely limits ISR on anything that's not a next.js optimized server-less platform. (which is not a realistic solution for all use cases)

**Comment by pekkai-solteq** [comment]
I actually just realized you can just `isrMemoryCahceSize` for disabling the cache altogether if you set the value to zero.

https://github.com/vercel/next.js/blob/canary/packages/next/server/config-shared.ts#L160
https://github.com/vercel/next.js/blob/b2acdbb788e48be4758f7555ae5fbe5ec5163eff/packages/next/server/next-server.ts#L298-L309
https://github.com/vercel/next.js/blob/canary/packages/next/server/incremental-cache.ts#L88-L97

Changes were introduced about a month since I had opened this discussion, I just had not looked back at this since.

https://github.com/vercel/next.js/commit/dbfce34ead5125ed9b0b952d50f7b0e1ae43154e#diff-aaa872d6d11669ea197d0696981330f0e98185690946f335a08326f52a004401

**Reply by aliehyaie** [reply]
I did this and my project is on 4 pods, but the problem still exists and has not been fixed.

**Comment by JordanTheriault** [comment]
This was pretty hard to find and there isn't much documentation on scaling with ISR. This is also true when using GCP's Cloud Run. Since scaling is important and many people don't deploy to Vercel, there should be better documentation on this.

**Comment by dottodot** [comment, marked as answer]
This update to the docs looks like a step in the right direction https://github.com/vercel/next.js/pull/36520/files#diff-be55212087c419559b8a4ef9dd38f625d18617ce76fa86ee3f9825c103d6fc24 

However it would be helpful if the docs had more details on how to self-host correctly. For example if mentions using a shared mount but is that for everything in .next or just some folders. I found putting everything on a shared mount resulted in module not found errors.

**Reply by gowthamhosur** [reply]
@dottodot did you happen to find out which folders go in shared mount, for self hosted apps to shared generated pages between pods?

**Reply by dottodot** [reply]
@gowthamhosur I believe it's .next/static, think I'm going to try a different approach using cloudfront and pushing .next/static to an s3 bucket. Not tried it yet but will try to remember to post here if I have any success.

**Reply by hannesl** [reply]
I did some experimentation with this, and it's not as easy as the docs say, I believe.

The `.next` directory is a mix of files that are generated at build time and at runtime. If you create a shared volume for the whole thing, the problem becomes migrating new files from the build into that volume. You will probably have to reset/swap out that volume on each deploy.

Generated pages are stored in `.next/server/pages` or `.next/server/app`. The problem is, again, that they're mixed with JS files from the build in the same directory structure. 🤷 

Two directories that should be OK to share and persist though are `.next/cache/fetch-cache` and `.next/cache/images`. That should be a good way to avoid duplicate requests against content backends.

Combine that with http caching in front of the app and it should be a quite scalable solution – although on-demand invalidation will not work.
